src/lib/ciul/checksum.c

```diff
@@ -295,6 +295,61 @@ uint32_t ef_tcp_checksum_ipx(int af, const void* ipx, const struct tcphdr* tcp,
     return ef_tcp_checksum((const struct iphdr*)ipx, tcp, iov, iovlen);
 }
 
+int ef_udp_checksum_is_correct(const struct iphdr* ip, const struct udphdr* udp,
+                               const struct iovec* iov, int iovlen)
+{
+  uint64_t csum64;
+  uint32_t csum;
+
+  csum64 = (uint64_t)ip->saddr + ip->daddr +    /* This is the UDP */
+           htons(IPPROTO_UDP) + udp->len;       /* pseudo-header */
+  csum64 = ip_csum64_partial(csum64, udp, sizeof(*udp));
+  csum64 = ip_csum64_partialv(csum64, iov, iovlen);
+  csum = ip_proto_csum64_finish(csum64);
+
+  return csum == 0;
```

Comment:
My checksum-fu isn't great, but I thought this was supposed to be `csum == 0 || csum == 0xffff`, i.e. ==+0 or ==-0. Can you test this code with a packet whose checksum is actually 0 (so the sender will have negated it)?

Reply:
Please check out the `src/tests/unit/lib/ciul/checksum.c`, which I've added.
The test checks `ef_tcp_checksum_is_correct()` (I think udp is the same) with a packet which tcp checksum is ffff, i.e, it is initially 0 but Linux inverts it (Onload would compute it as 0 for that packet).  The test shows that both 0 and ffff for that packet are treated as good checksum. So I guess that `csum == 0` is enough.

Reply:
If I change
```
  char tcpdata[] = {
    0xa9, 0xf6, 0x52, 0x13,
```
to
```
  char tcpdata[] = {
    0xa9, 0xf6, 0x52, 0x14,
```
then the test fails

Reply:
Right, yes, I think I've convinced myself that +0 (before negation) is unachievable - thanks for your patience. Ones' complement is mind-bending.